### ISSUE TITLE
Allow ocw-www author to set order of promos

### ISSUE THREAD

**Issue by pdpinch** [mitodl/ocw-hugo-projects#183]
As an ocw-www author, I'd like to be able to control the order of promos on the home page. 

Currently, the promos on the home page are displayed in alphabetical order by the promos' titles. 

Based on the discussion on this issue, it seems the preferred solution is to add a configuration option to collections to make them sortable. This would direct studio to present the items as a draggable list, similar to the menu UI. 

### related issues 

- allow ocw-www author to control which notification is displayed

### out of scope

- for this particular issue, we don't need to solve the question of how to sort across multiple pages of items in studio. We can address that in a follow up.

**Comment by Ferdi** [comment]
Is weight a numerical value that content author need to enter in studio ?  -- not that user friendly imo

**Comment by ChristopherChudzicki** [comment]
A sortable list seems like a natural UI in this scenario to me, but I think it would be significantly more effort to implement. If each promo WebsiteContent object has its own weight, then sorting Promp1 above promos above Promo2 ... PromoN might require updating all of their weights. To make the sortable list on the frontend, I think we'd need a new API (to get objects with nearby weight) and a new widget. (We have a sortable component, but it's only being used to sort data within the same WebsiteContent object, not across WebsiteContent objects.)

Is a simple numerical input field sufficient for an initial implementation, or do we want to think about the UI more?

**Comment by Ferdi** [comment]
>  Then, the active list would have to be chosen in site level metadata or something like that.

Not if it's of type "file" , cause it's just one of them.

Your idea of making collection sortable is a great one !

**Comment by gumaerc** [comment]
> Not if it's of type "file" , cause it's just one of them.

Right, that's an option.  That would require a theme modification to source the promos it renders from that file, but that would be trivial to do.  In fact, that would probably be the easiest way to do it.  Making the collection sortable would be more elegant imo, but the `ocw-studio` PR to add that functionality would require a bit more work.

**Comment by gumaerc** [comment]
Yea, That's the list we're thinking about allowing users to drag-order.  I do think it's the best approach, but a question that comes to mind is how to handle pagination.  Simply making a collection drag orderable in the case of Promos seems fairly simple, as there aren't that many of them.  However, a collection like Pages presents more of a problem if we were to allow this to be set on any collection.  The current UI only displays a certain amount of items and then paginates.  Also, I'm not sure how we would handle the search functionality when it comes to drag ordering.  Most OCW sites don't have a huge amount of pages, but some have hundreds so we would have to handle that case somehow.  Some kind of tab system to switch the UI might work.  Here's a rough mockup:

"Edit" mode:
![chrome_2022-06-10_16-44-23](https://user-images.githubusercontent.com/12089658/173147555-5ed12fb2-498c-4404-bb44-825e53f46409.png)

"Order" mode:
![image](https://user-images.githubusercontent.com/12089658/173149425-af9de0bb-183a-4fd1-8c80-000ee79f1365.png)

I just mocked this up in Photoshop and the menu ordering UI would obviously have to be tweaked for this, but we might be able to reuse it.

**Comment by pdpinch** [comment]
@Wassaf-Shahzad I saw [your video](https://user-images.githubusercontent.com/87968618/209152037-3fee66ba-a115-4a12-8615-a6892e4e4e7d.mov).

Can you use the same UI/UX as the menu section in Studio, like Carey's screenshots above (but without the settings icon and the trash icon)

**Comment by Wassaf-Shahzad** [comment]
@gumaerc @pdpinch I did some RnD and was able to make the site dashboard sortable but that was just a proof of concept. Currently the RelactionField (which houses SortableSelect) is used to make menu items sortable. I dont believe we can just plug this component into SiteDashboard and make it work. The main components responsible for the Drag and Drop are SortWrapper and SortableItem.

## Proposed Implementation
 - We can separate the Sorting functionality from SortableItem and create a new component SortableItemWrapper. (done)
 - This will insure the sorting functionality is not tied down to a specific UI design 
 - We might need to create a sortable version of the StudioListItemitem and StudioList components depending on the config.
 - The Backend Api will also need to change to store the order of  sortable collections.

**Comment by Ferdi** [comment]
Looking at the comments,  I'd suggest we go back to  using relation widget , as described by @gumaerc  above 

> The way we currently use the relation widget is by creating a new type of content, then defining a sortable relation widget within that new type of content. We would have to create a new type of content, something like "Promo Lists" and then define a sortable relation widget on that content. Then, the active list would have to be chosen in site level metadata or something like that

While making collection sortable sounds theoretically attractive , I don't see much practical use for it.  And it has some limitations:
-  It only make sense when nr of items in a collection is relatively small. 
-  You can only sort items once . E.g. with the relation widget approach one can create different lists with different sorting to be used in different context
- Does not seem straightforward to implement

**Comment by ChristopherChudzicki** [comment]
I agree with @Ferdi's points, and I agree that the relation widget is a good way to go.

@Ferdi I entirely agree with your second point ("You can only sort items once...") and think this is a good reason not to make collections sortable. I'd like to propose a simple solution for promos that only supports a single ordering, but if we ever needed multiple orderings in the future, we could make simple changes to support that.

What would you think of adding a new setting Settings menu to the... Something like "Homepage Settings":

```yaml
  - category: Settings
    label: Homepage
    name: homepage
    files:
      - file: config/_default/homepage.yaml
        name: homepage_settings
        label: Homepage Settings
        fields:
          - label: Featured Promos
            name: featured_promos
            widget: relation
            collection: promos
            display_field: title
            multiple: true
            sortable: true
```
which would result in
<img width="500" alt="Screenshot 2023-01-13 at 3 31 39 PM" src="https://user-images.githubusercontent.com/9010790/212416254-d151ee0f-8e5b-4e7a-a898-618a78ee6781.png">

@gumaerc I think this would be a good alternative to the "Promos Lists" idea mentioned above.

**Comment by Wassaf-Shahzad** [comment]
> I agree with @Ferdi's points, and I agree that the relation widget is a good way to go.
> 
> @Ferdi I entirely agree with your second point ("You can only sort items once...") and think this is a good reason not to make collections sortable. I'd like to propose a simple solution for promos that only supports a single ordering, but if we ever needed multiple orderings in the future, we could make simple changes to support that.
> 
> What would you think of adding a new setting Settings menu to the... Something like "Homepage Settings":
> 
> ```yaml
>   - category: Settings
>     label: Homepage
>     name: homepage
>     files:
>       - file: config/_default/homepage.yaml
>         name: homepage_settings
>         label: Homepage Settings
>         fields:
>           - label: Featured Promos
>             name: featured_promos
>             widget: relation
>             collection: promos
>             display_field: title
>             multiple: true
>             sortable: true
> ```
> 
> which would result in <img alt="Screenshot 2023-01-13 at 3 31 39 PM" width="500" src="https://user-images.githubusercontent.com/9010790/212416254-d151ee0f-8e5b-4e7a-a898-618a78ee6781.png">
> 
> I think that basically all this would require is:
> 
> * updating the ocw-www studio config file
> * a simple change to the homepage template in ocw-hugo-themes

@ChristopherChudzicki @pdpinch This is (probably )the best solution. It is far less complex and can save us future headaches. I can probably have a PR up and running hopefully tomorrow if this is approved.

**Comment by pdpinch** [comment]
I like this concept, I'm just not sure it belongs in the Settings menu, with the name "Homepage." Does "Promo List" or "Promo Collection" make sense?

**Comment by ChristopherChudzicki** [comment]
@pdpinch I think the best place we can put it right now is the Settings menu unless we want to create a new Collection type. 

Given that it under settings, what should its name be?

My 2c: If we want "Settings > TheNewItem"  to **only** control the order of the Promos, then I think "Promo List" or "Featured Promos" is a good title. If we think it might do more, then I "Settings > Homepage"

The reason I suggested "Settings > Homepage" is that I think of ...
- "Settings" controls settings about the site overall *like collaborators, or the header maybe, or perhaps some aspect of the search page*.
- "Settings > Homepage" controls settings about aspects of the homepage, https://ocw.mit.edu. This could include things like:
   - which promos are featured, and their order
   - which "Course List" feeds the featured courses on the homepage. *Currently the [filename is hardcoded](https://github.com/mitodl/ocw-hugo-themes/blob/main/www/layouts/partials/get_featured_courses.html#L3) as `featured-courses-homepage.md` in ocw-hugo-themes, but the filename could instead be read from a setting controlled under Settings > Homepage.*
    - which user stories are featured on homepage, and their order

**Comment by Wassaf-Shahzad** [comment]
@ChristopherChudzicki @pdpinch 
The following PRs are an implementation of the above mentions approach

#### Related PRs
- https://github.com/mitodl/ocw-hugo-themes/pull/1053
- https://github.com/mitodl/ocw-hugo-projects/pull/242

#### Thoughts
I just one though as that is regarding the path of homepage ( Name tbd ) setting file. Rather then storing in the default configs directory would it not be better to store it in the data directory of ocw-www content since it gives additional context about promos ?

**Comment by ChristopherChudzicki** [comment]
> Rather then storing in the default configs directory would it not be better to store it in the data directory of ocw-www content since it gives additional context about promos 

@Wassaf-Shahzad That makes sense to me. I think I just copied that from from Menus probably. (Should we store menus in data, too? Separate issue, though.)

**Comment by Wassaf-Shahzad** [comment]
> > Rather then storing in the default configs directory would it not be better to store it in the data directory of ocw-www content since it gives additional context about promos
> 
> @Wassaf-Shahzad That makes sense to me. I think I just copied that from from Menus probably. (Should we store menus in data, too? Separate issue, though.)

@ChristopherChudzicki No, since in case of menus Hugo just provides global access to data through `.Site.Menu` however if we try to store homepage.yml in `config/_defaults/` HUGO will just ignore it, Since its not one of  HUGOs predefined configurations (Params, Menus etc).

**Comment by Ferdi** [comment]
>  Rather then storing in the default configs directory would it not be better to store it in the data directory of ocw-www content since it gives additional context about promos

- Can you elaborate a bit more ?
- This seems like a config to me  so I would have thought it belongs under config/
- It's a yaml (homepge.yml ?) so it seems like a 'config'.   Aren't files under /data suppose to be json ?   @gumaerc

**Comment by Wassaf-Shahzad** [comment]
> > Rather then storing in the default configs directory would it not be better to store it in the data directory of ocw-www content since it gives additional context about promos
> 
> * Can you elaborate a bit more ?
> * This seems like a config to me  so I would have thought it belongs under config/
> * It's a yaml (homepge.yml ?) so it seems like a 'config'.   Aren't files under /data suppose to be json ?   @gumaerc

@Ferdi @pdpinch
- In cases of configuration directory, each file name represents a root configuration object such as `Params, Menus and Languages` [reference](https://gohugo.io/getting-started/configuration/#:~:text=Each%20file%20represents%20a%20configuration%20root%20object%2C%20such%20as%20params.toml%20for%20%5BParams%5D%2C%20menu(s).toml%20for%20%5BMenu%5D%2C%20languages.toml%20for%20%5BLanguages%5D%20etc%E2%80%A6). Hugo will not provide global level access for our custom config like it does for Menu because it is not Hugo's root configuration object.
- I agree with you that this is a configuration but I believe (I could be wrong) Hugo config folder is used to house configurations settings for Hugo itself such as any and all data we may need could be made available to use under .Site.Params by updating Params.yml.  [Reference to HUGO settings](https://gohugo.io/getting-started/configuration/#all-configuration-settings)
- Data folder can house .yml's as well. [reference](https://gohugo.io/templates/data-templates/#the-data-folder:~:text=These%20files%20must%20be%20YAML%2C%20JSON%2C%20XML%2C%20or%20TOML%20files%20(using%20the%20.yml%2C%20.yaml%2C%20.json%2C%20.xml%2C%20or%20.toml%20extension).%20The%20data%20will%20be%20accessible%20as%20a%20map%20in%20the%20.Site.Data%20variable.)

**Comment by gumaerc** [comment]
Yes, I wasn't so sure about this at the time I created `course.json` because all of the examples on the Hugo site using data templates were JSON. It's now clear though that anywhere where Hugo accepts structured data it accepts any of the standard formats; JSON, YAML, XML or TOML.

**Comment by Wassaf-Shahzad** [comment]
@pdpinch 
PRs related to this issue have been merged 
- https://github.com/mitodl/ocw-hugo-themes/pull/1053
- https://github.com/mitodl/ocw-hugo-projects/pull/242

**Comment by pdpinch** [comment]
I think https://github.com/mitodl/ocw-hugo-projects/pull/238 still needs to be reviewed and closed.

**Comment by ChristopherChudzicki** [comment]
Released in #243
